runtime/vm/jvminit.c

```diff
@@ -651,6 +658,29 @@ freeJavaVM(J9JavaVM * vm)
 		pool_kill(vm->cifArgumentTypesCache);
 		vm->cifArgumentTypesCache = NULL;
 	}
+
+	/* Clean up all resources created via allocateUpcallThunkMemory, including the generatered thunk
+	 * and the corresponding metadata of each entry stored in the hashtable.
+	 * See UpcallThunkMem.cpp for details.
+	 */
+	if (NULL != vm->thunkHeapWrapper) {
+		J9UpcallThunkHeapWrapper *thunkHeapWrapper = vm->thunkHeapWrapper;
+		J9PortVmemIdentifier vmemID = thunkHeapWrapper->vmemID;
+		UDATA byteAmount = j9vmem_supported_page_sizes()[0];
+		j9vmem_free_memory(vmemID.address, byteAmount, &vmemID);
+
+		if (NULL != thunkHeapWrapper->metaDataHashTable) {
+			J9HashTable *metaDataHashTable = thunkHeapWrapper->metaDataHashTable;
+			UDATA entryCount = hashTableGetCount(metaDataHashTable);
+			if (entryCount > 0) {
+				hashTableForEachDo(metaDataHashTable, (J9HashTableDoFn)freeUpcallMetaDataDoFn, NULL);
```

Comment:
the thunkheap is allocated out of the vmem, should the thunkheap be free'd (j9heap_free) first then the  j9vmem_free_memory ?

Reply:
Agreed and moved  `j9vmem_free_memory()` to the place together with `j9mem_free_memory(thunkHeapWrapper)` after releasing all thunk theaps.